packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4089,6 +4091,64 @@ class _DeleteTextAction<T extends DirectionalTextEditingIntent> extends ContextA
   bool get isActionEnabled => !state.widget.readOnly && state._value.selection.isValid;
 }
 
+/// An [Action] that transposes the characters immediately before and after the
+/// current collapsed selection.
+///
+/// When the cursor is at the end of the text, transposes the last two
+/// characters, if they exist.
+///
+/// When the cursor is at the start of the text, does nothing.
+class _TransposeCharactersAction extends ContextAction<TransposeCharactersIntent> {
+  _TransposeCharactersAction(this.state);
+
+  final EditableTextState state;
+
+  @override
+  Object? invoke(TransposeCharactersIntent intent, [BuildContext? context]) {
+    // Can't swap 1 or 0 characters.
+    if (state.textEditingValue.text.length <= 1) {
+      return null;
+    // Do nothing when there is no selection.
+    } else if (state.textEditingValue.selection == null) {
+      return null;
+    // Do nothing when the selection isn't collapsed.
+    } else if (!state.textEditingValue.selection.isCollapsed) {
+      return null;
+    // Do nothing when the selection is at the beginning of the text.
+    } else if (state.textEditingValue.selection.baseOffset == 0) {
+      return null;
+    }
+
+    final String text = state.textEditingValue.text;
+    final TextSelection selection = state.textEditingValue.selection;
+    final bool atEnd = selection.baseOffset == text.length;
+    final Characters before = selection.textBefore(text).characters;
+    final Characters after = selection.textAfter(text).characters;
+
+    final String replacement = atEnd
+        ? (before.takeLast(1) + before.characterAt(before.length - 2)).string
+        : after.first + before.last;
+
+    return Actions.invoke(
+      context!,
+      ReplaceTextIntent(
+        state.textEditingValue,
+        replacement,
+        TextRange(
+          start: atEnd
+              ? before.take(before.length - 2).string.length
+              : before.take(before.length - 1).string.length,
```

Comment:
what happens if `before` is empty?

Reply:
nit: also the `take` method is not lazy, so it creates a new string. Maybe use https://api.flutter.dev/flutter/characters/CharacterRange-class.html to calculate the start and end index.

Reply:
`before` can't be empty here because of all my else ifs up top.  I'll add an assertion for that.

Good call about `take`.  I've refactored this to use CharacterRange and lengths for everything.  I think it should perform much better, but let me know if I've missed anything.

I also added a test for extended grapheme clusters and surrogate pairs.

Reply:
Ah sorry you're right I thought you could transpose when the caret is placed at the start of the string.